### PR TITLE
Issue #3135: Indentation support for non-list child of block elements(if/else/for/while/switch/case)

### PR DESCRIPTION
**Pull request by Abhishek-kumar09**
Resolves Issue #3135
Diff Regression projects: https://gist.githubusercontent.com/Abhishek-kumar09/717190a7e2e06f4747b9c622692d4b99/raw/93fe7739cbdc52d7d17273a12d2b92328812c504/projects-to-test-on.properties
Diff Regression config: https://gist.githubusercontent.com/Abhishek-kumar09/0745ecf11934e0214b9187f9d4896e4a/raw/d52f23ee4f0db3d0b685d036c4321333a8e66ca5/my_check.xml

### CODE REVIEW

**Comment by Abhishek-kumar09** [comment]
[Diff report Tab Width 8](https://abhishek-kumar09.github.io/Issue3135/reportsTabWidth8/diff/index.html)

**Comment by romani** [comment]
@Abhishek-kumar09 , please rebase to resolve conflict.

**Comment by Abhishek-kumar09** [comment]
@romani Done
**New** [diff report](https://abhishek-kumar09.github.io/checkstylereport/reports/diff/index.html)

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/AbstractExpressionHandler.java [hunk 1]
```diff
@@ -414,7 +414,18 @@ protected final void checkExpressionSubtree(
...
+    }
+
+    /**
+     * Get the first ast for given expression.
+     *
+     * @param ast         the abstract syntax tree for which the starting ast is to be found
+     *
+     * @return the first ast of the expression
+     */
+    protected static DetailAST getFirstAstNode(DetailAST ast) {
```
Comment (Abhishek-kumar09) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/AbstractExpressionHandler.java:427]: This change is refractors in indentation code and doesn't much related to this Issue.
Should I make a new commit for this?

**Comment by Abhishek-kumar09** [comment]
Github, generate report

**Comment by Abhishek-kumar09** [comment]
Github, generate report

**Comment by romani** [comment]
Is it same problem in for/while/.. and other code ?
We should update their handlers in same way

**Comment by Abhishek-kumar09** [comment]
Actually it is same for `for` and `while` loop:
```
for(int i=0;i<8;i++)
int a =0; //warn
```
```

for(int i=0;i<8;i++)
System.out.println(); // no warn
```

It could detect the expression but not method calls and others. same for while.

I guess we only could place non list child with if, for and while.

**Comment by Abhishek-kumar09** [comment]
Github, generate report

**Comment by Abhishek-kumar09** [comment]
Github, generate report

**Comment by Abhishek-kumar09** [comment]
Github, generate report

**Comment by Abhishek-kumar09** [comment]
Github, generate report

**Comment by romani** [comment]
can you expalin https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/if_with_no_curly_2020215159/reports/diff/hibernate-orm/index.html#A8 ?
`'if' child has incorrect indentation level 15, expected level should be 16. | 180 | 16`
in real code position is 16, why we report it as 15 - https://github.com/CUBRID/hibernate-core/blob/master/hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/impl/BaseRegion.java#L215

**Comment by Abhishek-kumar09** [comment]
Yes I am on it, debugging.

**Comment by Abhishek-kumar09** [comment]
Github, generate report

**Comment by romani** [comment]
https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/if_with_no_curly_2020161423/reports/diff/guava/index.html#A1

Still problem with indentation in message va violation column

**Comment by Abhishek-kumar09** [comment]
@romani 
> https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/if_with_no_curly_2020161423/reports/diff/guava/index.html#A1
Still problem with indentation in message va violation column

I guess it is showing correct expected indent: 
this is how
basicOffset 4 with default indentation profile.

```
123   protected void runTestProfiled() throws Throwable { // indent:2 exp:4
124     long t0 = System.nanoTime(); //indent:4 exp:8
125     try {
126       super.runTest();
127     } finally {
128       long elapsedMillis = (System.nanoTime() - t0) / (1000L * 1000L); 
129       if (elapsedMillis >= profileThreshold) //indent:6 exp:8 + 4 = 12
130         System.out.printf("%n%s: %d%n", toString(), elapsedMillis); // indent:8 exp: 12+4 = 16 --> correct suggestion
131     }
132   }
```
I just have problem with these extra added/removed indents in checkstyle repo:
https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/if_with_no_curly_2020161423/reports/diff/checkstyle/index.html#A1

**Comment by Abhishek-kumar09** [comment]
> I just have problem with these extra added/removed indents in checkstyle repo:

Because When I check the two checkstyles locally they have the same error.

**Comment by Abhishek-kumar09** [comment]
> in real code position is 16, why we report it as 15 - https://github.com/CUBRID/hibernate-core/blob/master/hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/impl/BaseRegion.java#L215
in real code position is 16, why we report it as 15 ?

Sorry, I got your question late. 
In real, code position is 16 but the indentation level is always one less than the column number. 
e.g. Think of something placed at column 1, so we say that their indentation level is 0.
And hence something placed at 16 has15 as their indentation level.

**Comment by Abhishek-kumar09** [comment]
New Diff report: https://abhishek-kumar09.github.io/checkstyleDiffReportTempContribition/

**Comment by Abhishek-kumar09** [comment]
codecov showing covergae problem with AbstractExpressionHandler and BlockParentHandler, but on expanding it shows 100% covergae. How can I correct it.
image

**Comment by romani** [comment]
thanks a lot, yes it false alarm.
Indentation level is amount of spaces, but column statement beginning is always +1 (as we count columns from 1, as all IDEs and edits do).

@Abhishek-kumar09 , please remove all your old comment with old reports that are not relevant/actual any more.

**Comment by Abhishek-kumar09** [comment]
Github, generate report

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationIfElseWithNoCurly.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ * forceStrictCondition = false                                               //indent:1 exp:1
+ * lineWrappingIndentation = 4                                                //indent:1 exp:1
+ * tabWidth = 4                                                               //indent:1 exp:1
+ * throwsIndent = 4                                                           //indent:1 exp:1
+ *                                                                            //indent:1 exp:1
+ */                                                                           //indent:1 exp:1
+
+package com.puppycrawl.tools.checkstyle.checks.indentation.indentation; //indent:0 exp:0
+
+public class InputIndentationIfElseWithNoCurly { //indent:0 exp:0
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationIfElseWithNoCurly.java:16]: please copy this file few times and make "same" cases for `for/while/do-while/switch/case`

**Review by romani — CHANGES_REQUESTED** [review]
item:

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationIfElseWithNoCurly.java (hunk 2 above)
Comment (Abhishek-kumar09) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationIfElseWithNoCurly.java:16]: Done

Comment (Abhishek-kumar09) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationIfElseWithNoCurly.java:16]: made test cases for `for while and do-while`. switch is already there and had no issues like this with them.
They are already covered in there test cases, like: [InputIndentationInvalidSwitchIndent.java](https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/checks/indentation/indentation/InputIndentationInvalidSwitchIndent.java#L30-L46 )

**Review by romani — APPROVED** [review]

**Review by esilkensen — APPROVED** [review]

**Pull request merged**